README.md
# django-otp-sns
A django-otp plugin that sends and verifies tokens using AWS SNS.


CHANGES
v0.1.0 - August 21, 2019 - Initial Release
------------------------------------------

Initial release.


src/otp_sns/tests.py
from __future__ import absolute_import, division, print_function, unicode_literals

import re

from django.db import IntegrityError
from django.test.utils import override_settings

from django_otp.tests import TestCase

from .conf import settings


@override_settings(
    OTP_AMAZON_SNS_NO_DELIVERY=True,
    OTP_AMAZON_SNS_CHALLENGE_MESSAGE='{token}',
)
class TestAmazonSNS(TestCase):
    def setUp(self):
        try:
            self.alice = self.create_user('alice', 'password')
            self.bob = self.create_user('bob', 'password')
        except IntegrityError:
            self.skipTest("Unable to create test users.")
        else:
            self.alice.amazonsnsdevice_set.create(number='test',
                                                  key='01234567890123456789')
            self.bob.amazonsnsdevice_set.create(number='test',
                                                key='98765432109876543210')

        self._now = 1420099200
        self._delivered = None

    def test_instant(self):
        """ Verify a code the instant it was generated. """
        device = self.alice.amazonsnsdevice_set.get()
        with self.with_time(self._now):
            token = device.generate_challenge()
            ok = device.verify_token(token)

        self.assertTrue(ok)

    def test_default_key(self):
        device = self.alice.amazonsnsdevice_set.create(number='test')
        with self.with_time(self._now):
            token = device.generate_challenge()
            ok = device.verify_token(token)

        self.assertTrue(ok)

    def test_barely_made_it(self):
        """ Verify a code at the last possible second. """
        device = self.alice.amazonsnsdevice_set.get()
        with self.with_time(self._now):
            token = device.generate_challenge()
        with self.with_time(self._now + settings.OTP_AMAZON_SNS_TOKEN_VALIDITY):
            ok = device.verify_token(token)

        self.assertTrue(ok)

    def test_too_late(self):
        """ Try to verify a code one second after it expires. """
        device = self.alice.amazonsnsdevice_set.get()
        with self.with_time(self._now):
            token = device.generate_challenge()
        with self.with_time(self._now + settings.OTP_AMAZON_SNS_TOKEN_VALIDITY + 1):
            ok = device.verify_token(token)

        self.assertFalse(ok)

    def test_future(self):
        """ Try to verify a code from the future. """
        device = self.alice.amazonsnsdevice_set.get()
        with self.with_time(self._now + 1):
            token = device.generate_challenge()
        with self.with_time(self._now):
            ok = device.verify_token(token)

        self.assertFalse(ok)

    def test_code_reuse(self):
        """ Try to verify the same code twice. """
        device = self.alice.amazonsnsdevice_set.get()
        with self.with_time(self._now):
            token = device.generate_challenge()
            ok1 = device.verify_token(token)
            ok2 = device.verify_token(token)

        self.assertTrue(ok1)
        self.assertFalse(ok2)

    def test_cross_user(self):
        device = self.alice.amazonsnsdevice_set.get()
        with self.with_time(self._now):
            token = device.generate_challenge()
            ok = self.bob.amazonsnsdevice_set.get().verify_token(token)

        self.assertFalse(ok)

    @override_settings(
        OTP_AMAZON_SNS_NO_DELIVERY=False,
        OTP_AMAZON_SNS_TOKEN_TEMPLATE='Token is {token}',
    )
    def test_format(self):
        device = self.alice.amazonsnsdevice_set.get()
        with self._patch('otp_amazon_sns.models.AmazonSNSDevice._deliver_token', self._deliver_token):
            device.generate_challenge()
        match = re.match(r'^Token is (\d{6})$', self._delivered)

        self.assertIsNotNone(match)
        self.assertTrue(device.verify_token(match.group(1)))

    #
    # Utilities
    #

    def _deliver_token(self, token):
        self._delivered = token

    def with_time(self, timestamp):
        return self._patch('time.time', lambda: timestamp)

    def _patch(self, *args, **kwargs):
        try:
            from unittest import mock
        except ImportError:
            try:
                import mock
            except ImportError:
                self.skipTest("mock is not installed")

        return mock.patch(*args, **kwargs)


src/otp_sns/migrations/0001_initial.py
# -*- coding: utf-8 -*-
# Generated by Django 1.11.23 on 2019-08-21 13:34
from __future__ import unicode_literals

from django.conf import settings
from django.db import migrations, models
import django.db.models.deletion
import otp_sns.models

class Migration(migrations.Migration):

    initial = True

    dependencies = [
        migrations.swappable_dependency(settings.AUTH_USER_MODEL),
    ]

    operations = [
        migrations.CreateModel(
            name='AmazonSNSDevice',
            fields=[
                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
                ('name', models.CharField(help_text='The human-readable name of this device.', max_length=64)),
                ('confirmed', models.BooleanField(default=True, help_text='Is this device ready for use?')),
                ('number', models.CharField(help_text='The mobile number to deliver tokens to (E.164).', max_length=30)),
                ('key', models.CharField(default=otp_sns.models.default_key, help_text='A random key used to generate tokens (hex-encoded).', max_length=40, validators=[
                    otp_sns.models.key_validator])),
                ('last_t', models.BigIntegerField(default=-1, help_text='The t value of the latest verified token. The next token must be at a higher time step.')),
                ('user', models.ForeignKey(help_text='The user that this device belongs to.', on_delete=django.db.models.deletion.CASCADE, to=settings.AUTH_USER_MODEL)),
            ],
            options={
                'abstract': False,
                'verbose_name': 'Amazon SNS Device',
            },
        ),
    ]


src/otp_sns/models.py
from __future__ import absolute_import, division, print_function, unicode_literals

import logging
import time
from binascii import unhexlify

from boto3 import Session
from django.db import models
from django.utils.encoding import force_text
from django_otp.models import Device
from django_otp.oath import TOTP
from django_otp.util import random_hex, hex_validator

from .conf import settings

logger = logging.getLogger(__name__)


def default_key():
    return force_text(random_hex(20))


def key_validator(value):
    return hex_validator(20)(value)


class AmazonSNSDevice(Device):
    """
    A :class:`~django_otp.models.Device` that delivers codes via the Amazon SNS
    service. This uses TOTP to generate temporary tokens, which are valid for
    :setting:`OTP_AMAZON_SNS_TOKEN_VALIDITY` seconds. Once a given token has been
    accepted, it is no longer valid, nor is any other token generated at an
    earlier time.

    .. attribute:: number

        *CharField*: The mobile phone number to deliver to (E.164).

    .. attribute:: key

        *CharField*: The secret key used to generate TOTP tokens.

    .. attribute:: last_t

        *BigIntegerField*: The t value of the latest verified token.

    """
    number = models.CharField(
        max_length=30,
        help_text="The mobile number to deliver tokens to (E.164)."
    )

    key = models.CharField(
        max_length=40,
        validators=[key_validator],
        default=default_key,
        help_text="A random key used to generate tokens (hex-encoded)."
    )

    last_t = models.BigIntegerField(
        default=-1,
        help_text="The t value of the latest verified token. The next token must be at a higher time step."
    )

    class Meta(Device.Meta):
        verbose_name = "Amazon SNS Device"

    @property
    def bin_key(self):
        return unhexlify(self.key.encode())

    def generate_challenge(self):
        """
        Sends the current TOTP token to ``self.number``.

        :returns: :setting:`OTP_AMAZON_SNS_CHALLENGE_MESSAGE` on success.
        :raises: Exception if delivery fails.

        """
        totp = self.totp_obj()
        token = format(totp.token(), '06d')

        if hasattr(self, 'OTP_AMAZON_SNS_TOKEN_TEMPLATE'):
            # This allows upstream to set the token template on a per device
            # basis. An use case for this could be allowing every user's
            # organization to customize messages
            tpl = self.OTP_AMAZON_SNS_TOKEN_TEMPLATE
        else:
            tpl = settings.OTP_AMAZON_SNS_TOKEN_TEMPLATE

        message = tpl.format(token=token)
        if settings.OTP_AMAZON_SNS_NO_DELIVERY:
            logger.info(message)
        else:
            self._deliver_token(message)

        challenge = settings.OTP_AMAZON_SNS_CHALLENGE_MESSAGE.format(token=token)

        return challenge

    def _deliver_token(self, token):
        client = Session().client('sns')
        try:
            client.publish(
                PhoneNumber=self.number,
                Message=str(token)
            )
        except Exception as e:
            logger.exception('Error sending token by Amazon SNS: {0}'.format(e))
            raise

    def verify_token(self, token):
        try:
            token = int(token)
        except Exception:
            verified = False
        else:
            totp = self.totp_obj()
            tolerance = settings.OTP_AMAZON_SNS_TOKEN_VALIDITY

            for offset in range(-tolerance, 1):
                totp.drift = offset
                if (totp.t() > self.last_t) and (totp.token() == token):
                    self.last_t = totp.t()
                    self.save()

                    verified = True
                    break
            else:
                verified = False

        return verified

    def totp_obj(self):
        totp = TOTP(self.bin_key, step=1)
        totp.time = time.time()

        return totp
